src/codec.rs

```diff
@@ -145,61 +157,80 @@ impl<T: DeserializeOwned> Decoder for LanguageServerCodec<T> {
-                if let Some(header) = headers.iter().find(|h| h.name == "Content-Length") {
-                    match std::str::from_utf8(header.value) {
-                        Ok(content_len) => match content_len.parse() {
-                            // Successfully set `content_len` from the parsed "Content-Length"
-                            // value.
-                            Ok(content_len) => http_content_len = Some(content_len),
-                            // If there was an error parsing the "Content-Length" UTF-8 as a
-                            // `usize`, return the error.
-                            Err(_) => {
-                                src.advance(headers_len);
-                                return Err(ParseError::InvalidLength);
+        // If message length is known and source buffer doesn't contain the full message content yet
+        if src.len() < self.message_len.unwrap_or_default() {
+            return Ok(None);
+        }
+
+        let mut http_headers_err = Option::<ParseError>::default();
+
+        // If message length has not been parsed from "Content-Length" header yet
+        if self.message_len.is_none() {
+            // Placeholder used for parsing headers into
+            let dst = &mut [httparse::EMPTY_HEADER; 2];
+
+            // Parse the headers and try to extract values
+            match httparse::parse_headers(src, dst) {
+                // A complete set of headers was parsed succesfully
+                Ok(httparse::Status::Complete((headers_len, headers))) => {
+                    // Process the parsed headers
+                    for header in headers {
+                        match header.name {
+                            // Process a "Content-Length" header and extract the length value
+                            "Content-Length" => match std::str::from_utf8(header.value) {
+                                Ok(content_len) => match content_len.parse::<usize>() {
+                                    Ok(content_len) => {
+                                        self.message_len = Some(content_len);
+                                    }
+                                    Err(err) => {
+                                        http_headers_err =
+                                            Some(ParseError::InvalidContentLength(err));
+                                        break;
+                                    }
+                                },
+                                Err(err) => {
+                                    http_headers_err = Some(ParseError::Utf8(err));
+                                }
+                            },
+                            // Process a "Content-Type" header and just check that the value is of an expected value
+                            "Content-Type" => {
+                                if header.value != b"application/vscode-jsonrpc; charset=utf-8" {
```

Comment:
FYI, according to [the LSP specification](https://microsoft.github.io/language-server-protocol/specification#contentPart), the only supported `charset` values are `utf-8` or `utf8` (for backwards compatibility only). I feel pretty confident we should return an error on invalid `charset` values, but I think this fix is orthogonal to this PR and #314. I don't want to hold up this PR from getting merged!

Reply:
That seems reasonable to me. I can add a check for that in a new PR.

Reply:
No worries! This should already be rolled into mainline in [9f701d7](https://github.com/ebkalderon/tower-lsp/commit/9f701d7d89eccee4900f880cf8888942336fb5eb).